src/Control/Monad/Schedule/FreeAsync.hs

```diff
@@ -0,0 +1,92 @@
+
+-- monad-schedule
+import Control.Monad.Schedule.Class (MonadSchedule (..))
+import Control.Monad.Trans.Class
+
+type FreeAsync = FreeAsyncT IO
+newtype FreeAsyncT m a = FreeAsyncT {getFreeAsyncT :: ProgramT MVar m a}
+  deriving newtype (Functor, Applicative, Monad, MonadTrans)
+
+instance (MonadIO m) => MonadIO (FreeAsyncT m) where
+  liftIO action = FreeAsyncT $ do
+    var <- liftIO newEmptyMVar
+    liftIO $ forkIO $ putMVar var =<< action
+    singleton var
+
+runFreeAsyncT :: (MonadIO m) => FreeAsyncT m a -> m a
+runFreeAsyncT = interpretWithMonadT (liftIO . takeMVar) . getFreeAsyncT
+
+runFreeAsync :: FreeAsync a -> IO a
+runFreeAsync = runFreeAsyncT
+
+asyncMVar :: MVar a -> FreeAsyncT m a
+asyncMVar = FreeAsyncT . singleton
+
+data MVarCont m a = forall b.
+  MVarCont
+  { mvar :: MVar b
+  , cont :: b -> ProgramT MVar m a
+  }
+
+embedMVarCont :: (Monad m) => MVarCont m a -> FreeAsyncT m a
+embedMVarCont MVarCont {mvar, cont} = FreeAsyncT $ unviewT $ mvar :>>= cont
+
+instance (MonadIO m) => MonadSchedule (FreeAsyncT m) where
+  schedule actions = retryForever $ getFreeAsyncT <$> actions
+    where
+      retryForever :: (MonadIO m) => NonEmpty (ProgramT MVar m a) -> FreeAsyncT m (NonEmpty a, [FreeAsyncT m a])
+      retryForever actions = do
+        views <- lift (mapM viewT actions)
+        case partitionNonEmpty $ viewToEither <$> views of
+          Left (as, conts) -> return (as, embedMVarCont <$> conts)
+          Right conts -> do
+            (progressed, conts) <- lift $ retryProgresses conts
+            retryForever $ appendList progressed $ getFreeAsyncT . embedMVarCont <$> conts
+
+      viewToEither :: ProgramViewT MVar m a -> Either a (MVarCont m a)
+      viewToEither (Return a) = Left a
+      viewToEither (mvar :>>= cont) = Right MVarCont {mvar, cont}
+
+      partitionNonEmpty :: NonEmpty (Either a b) -> Either (NonEmpty a, [b]) (NonEmpty b)
+      partitionNonEmpty (Left a :| abs) = let (as, bs) = partitionEithers abs in Left (a :| as, bs)
+      partitionNonEmpty (Right b :| abs) = case partitionEithers abs of
+        ([], bs) -> Right $ b :| bs
+        (a : as, bs) -> Left (a :| as, b : bs)
+
+      tryProgress :: (MonadIO m) => MVarCont m a -> m (Either (ProgramT MVar m a) (MVarCont m a))
+      tryProgress mvarcont@MVarCont {mvar, cont} = do
+        result <- liftIO $ tryTakeMVar mvar
+        return $ maybe (Right mvarcont) (Left . cont) result
+
+      tryProgresses :: (MonadIO m) => NonEmpty (MVarCont m a) -> m (Either (NonEmpty (ProgramT MVar m a), [MVarCont m a]) (NonEmpty (MVarCont m a)))
+      tryProgresses conts = partitionNonEmpty <$> mapM tryProgress conts
+
+      retryProgresses :: (MonadIO m) => NonEmpty (MVarCont m a) -> m (NonEmpty (ProgramT MVar m a), [MVarCont m a])
+      retryProgresses conts = do
+        result <- tryProgresses conts
+        case result of
+          Left progress -> return progress
+          Right _ -> do
+            liftIO $ yield >> threadDelay 100
```

Comment:
Maybe these 100 should be tunable? E.g. as an argument to `runFreeAsync` or as a type level argument?

Reply:
Let's leave it for now.